### PR TITLE
Introduce warnings for nonReserved words moving to Reserved

### PR DESCRIPTION
**Pull request by sujay-jain**
Fix for https://github.com/prestodb/presto/issues/13586 
Open to suggestions for a better way to keep the set of the words. Opted for a static set due to the temporary nature of the warning, and the assumption that this list would not change often. Downside being that this approach is not very extendible. 

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java [hunk 1]
```diff
@@ -234,6 +234,11 @@ public void exitNonReserved(SqlBaseParser.NonReservedContext context)
             if (token.getText().equalsIgnoreCase("CURRENT_ROLE")) {
```
Comment (highker) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java]: put `CURRENT_ROLE` in `RESERVED_WORDS_WARNING` as well

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java [hunk 2]
```diff
@@ -13,14 +13,21 @@
...
+import java.util.Set;
 
 import static java.util.Objects.requireNonNull;
 
 public class SqlParserOptions
 {
+    //TODO: discuss - not the most extendible way of maintaining such a list.
+    // Considerations: this list would change infrequently, and the warning is temporary.
+    protected static final Set<String> RESERVED_WORDS_WARNING = ImmutableSet.of(
+            "CALLED", "called", "DETERMINISTIC", "deterministic", "FUNCTION", "function", "LANGUAGE", "language", "RETURN", "return", "RETURNS", "returns", "SQL", "sql");
```
Comment (highker) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java]: upper case is good enough. Use `equalsIgnoreCase` to check

Comment (sujay-jain) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java]: Since this is a Set, it doesn't have ```containsIgnoreCase```. I could use upper case and then do something like ```set.contains(word.toUpperCase())``` as an alternative, but the tradeoff is that we would be paying the cost of ```toUpperCase``` everytime. I agree that would be cleaner but with a tiny performance overhead :)

Comment (highker) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java]: This is more about correctness than performance. `DeTeRMINIstic` is also a valid reserved word.

Comment (sujay-jain) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java]: good point. I didn't think of that :D

Comment (rongrong) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java]: You can also use `toUpperCase` or `toLowerCase` for comparison. As @highker said, this is about correctness. SQL keywords are not case sensitive.

File: presto-tests/src/test/java/com/facebook/presto/execution/TestWarnings.java [hunk 3]
```diff
@@ -83,6 +83,17 @@ public void testNonReservedWordWarning()
         assertWarnings(queryRunner, session, query, ImmutableList.of(PARSER_WARNING.toWarningCode()));
     }
 
+    @Test
+    public void testNewReservedWordsWarning()
```
Comment (rongrong) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestWarnings.java]: Also add a test that quoted identifier does not trigger warning.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java [hunk 4]
```diff
@@ -231,11 +231,7 @@ public void exitNonReserved(SqlBaseParser.NonReservedContext context)
             context.getParent().removeLastChild();
 
             Token token = (Token) context.getChild(0).getPayload();
-            if (token.getText().equalsIgnoreCase("CURRENT_ROLE")) {
-                warningConsumer.accept(new ParsingWarning(format("Reserved word used: %s", token.getText()), token.getLine(), token.getCharPositionInLine()));
-            }
-
-            if (SqlParserOptions.RESERVED_WORDS_WARNING.contains(token.getText())) {
+            if (SqlParserOptions.RESERVED_WORDS_WARNING.contains(token.getText().toUpperCase())) {
```
Comment (rongrong) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java]: We structure commits logically, which means that this bug fix should belong to the first commit (if you want to keep 2 separate commits). Since adding `CURRENT_ROLE` to `RESERVED_WORDS_WARNING` is very much related and very short you can merge the two commits into one. Please do `git push -f` to update the PR after you restructure your commits. Thanks!

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java [hunk 5]
```diff
@@ -26,7 +26,7 @@
     //TODO: discuss - not the most extendible way of maintaining such a list.
     // Considerations: this list would change infrequently, and the warning is temporary.
     protected static final Set<String> RESERVED_WORDS_WARNING = ImmutableSet.of(
-            "CALLED", "called", "DETERMINISTIC", "deterministic", "FUNCTION", "function", "LANGUAGE", "language", "RETURN", "return", "RETURNS", "returns", "SQL", "sql");
+            "CALLED", "CURRENT_ROLE", "DETERMINISTIC", "FUNCTION", "LANGUAGE", "RETURN", "RETURNS", "SQL");
```
Comment (rongrong) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java]: Same here, most of this change belongs to the first commit (other than adding `CURRENT_ROLE`).

**Comment by rongrong** [comment]
Please sign the EasyCLA(click on Details).  before you update the PR so authorization check would pass. Thanks!

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java [hunk 6]
```diff
@@ -231,9 +231,10 @@ public void exitNonReserved(SqlBaseParser.NonReservedContext context)
             context.getParent().removeLastChild();
 
             Token token = (Token) context.getChild(0).getPayload();
-            if (token.getText().equalsIgnoreCase("CURRENT_ROLE")) {
-                warningConsumer.accept(new ParsingWarning(format("Reserved word used: %s", token.getText()), token.getLine(), token.getCharPositionInLine()));
+            if (SqlParserOptions.RESERVED_WORDS_WARNING.contains(token.getText().toUpperCase())) {
```
Comment (caithagoras) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java]: static import `RESERVED_WORDS_WARNING`

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java [hunk 7]
```diff
@@ -13,14 +13,21 @@
...
 import com.google.common.collect.Iterables;
 
 import java.util.EnumSet;
+import java.util.Set;
 
 import static java.util.Objects.requireNonNull;
 
 public class SqlParserOptions
 {
+    //TODO: discuss - not the most extendible way of maintaining such a list.
```
Comment (caithagoras) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java]: nit: space after `//`

File: presto-tests/src/test/java/com/facebook/presto/execution/TestWarnings.java (hunk 3 above)
Comment (sujay-jain) [reply presto-tests/src/test/java/com/facebook/presto/execution/TestWarnings.java]: thanks for catching this! Added a test that verifies no warnings for the case of ```"CALLED"``` as an example

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java (hunk 4 above)
Comment (sujay-jain) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java]: thanks for explaining this :) I've squashed all my commits into one with the commit message most relevant to the issue. Happy to make more changes if something seems off

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java [hunk 8]
```diff
@@ -13,14 +13,21 @@
...
 import com.google.common.collect.Iterables;
 
 import java.util.EnumSet;
+import java.util.Set;
 
 import static java.util.Objects.requireNonNull;
 
 public class SqlParserOptions
 {
+    // TODO: discuss - not the most extendible way of maintaining such a list.
```
Comment (rongrong) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParserOptions.java]: Is this still a TODO? If not please remove it. Thanks!

File: presto-tests/src/test/java/com/facebook/presto/execution/TestWarnings.java [hunk 9]
```diff
@@ -83,6 +83,28 @@ public void testNonReservedWordWarning()
...
+    {
+        String query = "SELECT CALLED, t.called FROM (VALUES (3)) t(called)";
+        Session session = testSessionBuilder()
```
Comment (rongrong) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestWarnings.java]: You can directly use `TEST_SESSION`

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java [hunk 10]
```diff
@@ -231,9 +232,10 @@ public void exitNonReserved(SqlBaseParser.NonReservedContext context)
             context.getParent().removeLastChild();
 
             Token token = (Token) context.getChild(0).getPayload();
-            if (token.getText().equalsIgnoreCase("CURRENT_ROLE")) {
-                warningConsumer.accept(new ParsingWarning(format("Reserved word used: %s", token.getText()), token.getLine(), token.getCharPositionInLine()));
+            if (RESERVED_WORDS_WARNING.contains(token.getText().toUpperCase())) {
+                warningConsumer.accept(new ParsingWarning(format("This will soon be made a reserved word: %s", token.getText()), token.getLine(), token.getCharPositionInLine()));
```
Comment (rongrong) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java]: `"%s should be a reserved word, please use double quote (\"%s\"). This will be made a reserved word in future release."`

Comment (sujay-jain) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/SqlParser.java]: thanks!
